Rename Note Type: randomizeName if it is already exist

Akshay0701 opened the pull request:
## Purpose / Description
In ankidroid whenever we rename with pre-existed note type then it gets saved with duplicate names 
whereas in AnkiDesktop it appends randomize keys to renamed string.


## Output
<img src="https://user-images.githubusercontent.com/52353967/128861276-557e612c-bec0-4f6c-9653-bda876b17908.gif" width="250" height="500" />

Arthur-Milchior commented:
Would you mind splitting in two commits. One which create the function and move code in it, the other that do the actual functional change. It will be easier to check

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/ModelBrowser.java:
```diff
@@ -452,6 +471,11 @@ private void renameModelDialog() {
                                             .replaceAll("[\"\\n\\r]", "");
+
+                                    if (mExistingModelsNames.contains(deckName)) {
+                                        deckName = randomizeName(deckName);
```
`randomizeName` is likely incorrect. I would expect that the following should be ported to Models.java and used instead

https://github.com/ankidroid/Anki-Android/blob/ee7cd786ca991f23f8df17879816e80712fefe31/AnkiDroid/src/main/java/com/ichi2/libanki/ModelsV16.kt#L271-L281

Akshay0701 replied:
``` java
     /*
     * Takes current timestamp from col
     * Used to append to the end of new note types to dissuade
     * User from reusing names (which are technically not unique however
     */
    private String randomizeName(String s) {
        return s + " " + checksum(String.valueOf(getCol().getTime().intTimeMS())).substring(0, 5);
    }
```

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/ModelBrowser.java:
```diff
@@ -73,6 +73,8 @@
     private ArrayList<Long> mModelIds;
     private ArrayList<DisplayPair> mModelDisplayList;
+    private ArrayList<String> mNewModelLabels;
+    private ArrayList<String> mExistingModelsNames;
```
mExistingModelNames

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/ModelBrowser.java:
```diff
@@ -400,6 +379,41 @@ private void addNewNoteType(String modelName, int position) {
+            mNewModelNames.add(defaultName);
+        }
+
+        if (mModels != null) {
```
Can `mModels` be null here? It doesn't feel possible

david-allison-1 commented on AnkiDroid/src/main/java/com/ichi2/anki/ModelBrowser.java:
```diff
@@ -302,36 +305,12 @@ private void updateSubtitleText() {
-                existingModelsNames.add(name);
-            }
+        if (mExistingModelsNames == null || mNewModelLabels == null) {
+            return;
```
When will this be re-called if the values are null?

Akshay0701 replied:
I am returning method call if values are null and not recalling it

Akshay0701 replied:
I think this value will never become null, I should remove this if statement.

david-allison-1 reviewed (commented):
Could you split out the refactoring commit further. I don't understand how it works logically in its current form

Akshay0701 commented:
> I don't understand how it works logically in its current form and it feels like a functional change

My mistake,

now I arranged commits in a more structured way,

first I fixed the randomizeName method (as David suggested),

then I extracted note type initialization code to a method
(because we going to need to call that in the `renameModelDialog` function).

and lastly, I added if statement in the renameModelDialog method to check if renamed string already exists or not.


let me know if there is any confusion left....

david-allison-1 reviewed (approved):
LGTM.

A unit test here would be great, as we're likely going to modify this functionality for the V16 conversion

mikehardy reviewed (approved):
Also LGTM - going to merge w/o test 🙈

mikehardy commented:
This one took a conflict from #9286 / #9343 I believe - I think I have fixed them and this can squash merge if it passes CI

mikehardy commented:
Hi there! **September OpenCollective Notice**

Apologies for the delay, and if you're already submitted something to OpenCollective for PRs merged in **September** 2021, you may feel free to ignore this and it will be processed shortly

Just a friendly notice that we try to process OpenCollective payments monthly - it's time for September 2021 submissions

If you are interested in compensation for this work, the process with details is here:

https://github.com/ankidroid/Anki-Android/wiki/OpenCollective-Payment-Process#how-to-get-paid

(I only post one comment per person to avoid spamming you, regardless of the number of PRs merged, but this note applies to all PRs merged for in the month of August)

Thanks!
